src/porepy/models/run_convergence_study.py

```diff
@@ -0,0 +1,217 @@
+import os
+import logging
+from typing import (  # noqa
+    Any,
+    Callable,
+    Coroutine,
+    Generator,
+    Generic,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+from pathlib import Path
+from pprint import pformat
+import functools
+import inspect
+
+import pendulum
+import porepy as pp
+
+from porepy.models.contact_mechanics_model import ContactMechanics
+from porepy.utils.grid_refinement import gb_coarse_fine_cell_mapping, refine_mesh
+from porepy.utils.grid_convergence import grid_error
+
+
+logger = logging.getLogger(__name__)
+
+
+def gb_refinements(
+        network: Union[pp.FractureNetwork3d, pp.FractureNetwork2d],
+        gmsh_folder_path: Union[str, Path],
+        mesh_args: dict,
+        n_refinements: int = 0
+) -> List[pp.GridBucket]:
```

Comment:
Could you review my new suggestion (see 3c826a6), using generators?

At this point, `refine_mesh_by_splitting` should be optimized. I have not done that for `run_model_for_convergence_study` yet (since it needs the finest grid anyways).